src/lib/block_producer/block_producer.ml

```diff
@@ -679,3 +690,243 @@ let run ~logger ~prover ~verifier ~trust_system ~get_completed_work
+          let%bind breadcrumb =
+            Breadcrumb.build ~logger ~precomputed_values ~verifier
+              ~trust_system ~parent:crumb ~transition ~sender:None
+            |> Deferred.Result.map_error ~f:(function
+                 | `Invalid_staged_ledger_diff e ->
+                     `Invalid_staged_ledger_diff (e, staged_ledger_diff)
+                 | ( `Fatal_error _
+                   | `Invalid_genesis_protocol_state
+                   | `Invalid_staged_ledger_hash _
+                   | `Not_selected_over_frontier_root
+                   | `Parent_missing_from_frontier ) as err ->
+                     err )
+          in
+          [%str_log trace]
+            ~metadata:[("breadcrumb", Breadcrumb.to_yojson breadcrumb)]
+            Block_produced ;
+          let metadata =
+            [("state_hash", State_hash.to_yojson transition_hash)]
+          in
+          Coda_metrics.(Counter.inc_one Block_producer.blocks_produced) ;
+          let%bind.Async () =
+            Strict_pipe.Writer.write transition_writer breadcrumb
+          in
+          [%log debug] ~metadata
+            "Waiting for block $state_hash to be inserted into frontier" ;
+          Deferred.choose
+            [ Deferred.choice
+                (Transition_registry.register transition_registry
+                   transition_hash)
+                (Fn.const (Ok `Transition_accepted))
+            ; Deferred.choice
+                ( Time.Timeout.create time_controller (Time.Span.of_ms 20000L)
+                    ~f:(Fn.const ())
+                |> Time.Timeout.to_deferred )
+                (Fn.const (Ok `Timed_out)) ]
+          >>= function
+          | `Transition_accepted ->
+              [%log info] ~metadata
+                "Generated transition $state_hash was accepted into \
+                 transition frontier" ;
+              return ()
+          | `Timed_out ->
+              (* FIXME #3167: this should be fatal, and more importantly,
+                 shouldn't happen.
+              *)
+              [%log fatal] ~metadata
+                "Timed out waiting for generated transition $state_hash to \
+                 enter transition frontier. Continuing to produce new blocks \
+                 anyway. This may mean your CPU is overloaded. Consider \
+                 disabling `-run-snark-worker` if it's configured." ;
+              return ()
+        in
+        let transition_error_msg_prefix = "Validation failed: " in
+        let transition_reason_for_failure =
+          " One possible reason could be a ledger-catchup is triggered before \
+           we produce a proof for the produced transition."
+        in
+        let exn_breadcrumb name =
+          raise
+            (Error.to_exn
+               (Error.of_string
+                  (sprintf
+                     "Error building breadcrumb from produced transition: %s"
+                     name)))
+        in
+        match%bind emit_breadcrumb () with
```

Comment:
This code seems very similar to what's in the existing `run`.

Can the commonality be factored out?